### PR TITLE
Imperative Resnet

### PR DESCRIPTION
**Pull request by velconia**
1. Fix the bug of static BatchNorm layer

2. Add the implementation of BatchNorm in imperative mode

3. Add ResNet unit test for imperative mode

4. Support single GPU card

### CODE REVIEW

File: paddle/fluid/imperative/layer.cc [hunk 1]
```diff
@@ -34,22 +38,68 @@ std::map<int, py::object> py_funcs_;
...
+class TensorAddToFunctor : public boost::static_visitor<> {
+ public:
+  TensorAddToFunctor(int64_t numel, const T* x, T* y)
+      : numel_(numel), x_(x), y_(y) {}
+
+  void operator()(const platform::CPUPlace& place) {
+    platform::CPUDeviceContext* ctx = dynamic_cast<platform::CPUDeviceContext*>(
+        platform::DeviceContextPool::Instance().Get(place));
+    auto blas =
+        operators::math::GetBlas<platform::CPUDeviceContext, float>(*ctx);
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.cc]: float->T?

File: paddle/fluid/imperative/layer.cc [hunk 2]
```diff
@@ -34,22 +38,68 @@ std::map<int, py::object> py_funcs_;
...
+    blas.AXPY(numel_, 1., x_, y_);
+  }
+
+#ifdef PADDLE_WITH_CUDA
+  void operator()(const platform::CUDAPlace& place) {
+    platform::CUDADeviceContext* ctx =
+        dynamic_cast<platform::CUDADeviceContext*>(
+            platform::DeviceContextPool::Instance().Get(place));
+    auto blas =
+        operators::math::GetBlas<platform::CUDADeviceContext, float>(*ctx);
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.cc]: float->T?

File: paddle/fluid/imperative/layer.cc [hunk 3]
```diff
@@ -34,22 +38,68 @@ std::map<int, py::object> py_funcs_;
...
+    auto blas =
+        operators::math::GetBlas<platform::CPUDeviceContext, float>(*ctx);
+    blas.AXPY(numel_, 1., x_, y_);
+  }
+
+#ifdef PADDLE_WITH_CUDA
+  void operator()(const platform::CUDAPlace& place) {
+    platform::CUDADeviceContext* ctx =
+        dynamic_cast<platform::CUDADeviceContext*>(
+            platform::DeviceContextPool::Instance().Get(place));
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.cc:60]: make DeviceContext a template? Seems to work in some kernels.

File: paddle/fluid/imperative/layer.cc [hunk 4]
```diff
@@ -34,22 +38,68 @@ std::map<int, py::object> py_funcs_;
...
 
-void AddTo(Variable* src, Variable* dst) {
-  framework::LoDTensor* dst_tensor = dst->GetMutable<framework::LoDTensor>();
-  framework::LoDTensor* src_tensor = src->GetMutable<framework::LoDTensor>();
+namespace detail {
+
+template <typename T>
+class TensorAddToFunctor : public boost::static_visitor<> {
+ public:
+  TensorAddToFunctor(int64_t numel, const T* x, T* y)
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.cc:46]: Do we have this function already somewhere? @chengduoZH

File: paddle/fluid/imperative/layer.cc [hunk 5]
```diff
@@ -34,22 +38,68 @@ std::map<int, py::object> py_funcs_;
...
+
+ private:
+  int64_t numel_;
+  const T* x_;
+  T* y_;
+};
+
+}  // namespace detail
+
+void AddGradTo(Variable* src, Variable* dst, platform::Place place) {
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.cc]: why change the name, seems AddTo is fine.

File: paddle/fluid/pybind/pybind.cc [hunk 6]
```diff
@@ -137,6 +137,9 @@ PYBIND11_MODULE(core, m) {
       .def("_grad_ivar",
            [](const imperative::VarBase &self) { return self.grads_; },
            py::return_value_policy::reference)
+      .def("_cpu_tensor",
```
Comment (panyx0718) [inline paddle/fluid/pybind/pybind.cc]: Can we design API like this:
```
var.to(place)

var.to(fluid.CPUPlace())
var.to(fluid.CUDAPlacae(0))
```

File: paddle/fluid/imperative/layer.cc [hunk 7]
```diff
@@ -188,8 +261,10 @@ void VarBase::RunBackward() {
 
   VLOG(3) << "start backward";
   auto grads_t = grads_->var_->GetMutable<framework::LoDTensor>();
-  float* data = grads_t->mutable_data<float>(platform::CPUPlace());
-  std::fill(data, data + grads_t->numel(), 1.0);
+  operators::math::set_constant(
+      *(platform::DeviceContextPool::Instance().Get(
+          var_->GetMutable<framework::LoDTensor>()->place())),
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.cc:270]: grads_t?

File: paddle/fluid/imperative/layer.h [hunk 8]
```diff
@@ -148,6 +153,8 @@ class VarBase {
 
   framework::LoDTensor& GradValue();
 
+  framework::LoDTensor* CopiedTensor() const;
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.h]: no a good name.

File: paddle/fluid/imperative/layer.h [hunk 9]
```diff
@@ -176,7 +183,8 @@ class OpBase {
       : op_desc_(nullptr),
         forward_id_(-1),
         grad_op_desc_(nullptr),
-        backward_id_(-1) {}
+        backward_id_(-1),
+        expected_place_(platform::CPUPlace()) {}
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.h]: just call place_?

File: paddle/fluid/imperative/tracer.cc [hunk 10]
```diff
@@ -31,16 +35,38 @@ void CreateGradOp(const framework::OpDesc& op_desc,
...
+  PADDLE_ENFORCE_NOT_NULL(dev_ctx,
+                          "Could not get valid device from forward op");
   auto& var_t = var->Get<framework::LoDTensor>();
-  float* data =
-      grad_var->GetMutable<framework::LoDTensor>()->mutable_data<float>(
-          var_t.dims(), platform::CPUPlace());
-  std::fill(data, data + var_t.numel(), 0.0);
+  grad_var->GetMutable<framework::LoDTensor>()->mutable_data<float>(
+      var_t.dims(), dev_ctx->GetPlace());
+  operators::math::set_constant(
```
Comment (panyx0718) [inline paddle/fluid/imperative/tracer.cc:45]: var_t?

File: python/paddle/fluid/framework.py [hunk 11]
```diff
@@ -2321,9 +2326,16 @@ def _get_var(name, program=None):
 
 
 @contextlib.contextmanager
-def _imperative_guard(tracer):
+def _imperative_guard(tracer, place):
```
Comment (panyx0718) [inline python/paddle/fluid/framework.py]: don't do this now.
just infer the place from input vars.

File: python/paddle/fluid/imperative/base.py [hunk 12]
```diff
@@ -25,18 +25,29 @@ def enabled():
 
 
 @contextlib.contextmanager
-def guard():
+def guard(device=0):
```
Comment (panyx0718) [inline python/paddle/fluid/imperative/base.py]: remove this API change. infer from input tensor

File: paddle/fluid/imperative/layer.cc (hunk 1 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.cc]: Done

File: paddle/fluid/imperative/layer.cc (hunk 2 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.cc]: Done

File: paddle/fluid/imperative/layer.cc (hunk 5 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.cc]: Done

File: paddle/fluid/imperative/layer.cc (hunk 4 above)
Comment (chengduoZH) [reply paddle/fluid/imperative/layer.cc:46]: For tensor `add`, we usually use Eigen or `elementwise_add`.

File: python/paddle/fluid/imperative/base.py (hunk 12 above)
Comment (velconia) [reply python/paddle/fluid/imperative/base.py]: Without this API, we should have a way to make parameters be put in GPU

File: paddle/fluid/imperative/layer.h (hunk 9 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.h]: done

File: paddle/fluid/imperative/layer.h (hunk 8 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.h]: done

File: paddle/fluid/pybind/pybind.cc (hunk 6 above)
Comment (velconia) [reply paddle/fluid/pybind/pybind.cc]: done

File: python/paddle/fluid/imperative/base.py (hunk 12 above)
Comment (velconia) [reply python/paddle/fluid/imperative/base.py]: And a way to make op without input keep in GPU, e.g.. ```fill_constant```

File: paddle/fluid/imperative/layer.cc [hunk 13]
```diff
@@ -162,7 +240,7 @@ std::map<std::string, std::vector<VarBase*>> OpBase::ApplyGrad() {
     PADDLE_ENFORCE_NOT_NULL(op_kernel, "only support op with kernel");
 
     framework::Scope scope;
-    platform::CPUPlace place;
+    platform::Place place = place_;
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.cc]: why copy here?

File: python/paddle/fluid/tests/unittests/test_imperative_resnet.py [hunk 14]
```diff
@@ -0,0 +1,533 @@
...
+        batch_size = train_parameters["batch_size"]
+        batch_num = 1
+        with fluid.imperative.guard():
+            fluid.default_startup_program().random_seed = seed
+            fluid.default_main_program().random_seed = seed
+
+            resnet = ResNet()
+            optimizer = optimizer_setting(train_parameters)
+            np.random.seed(seed)
+            import random
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_imperative_resnet.py:219]: don't do this anymore...

File: python/paddle/fluid/tests/unittests/test_imperative_resnet.py [hunk 15]
```diff
@@ -0,0 +1,533 @@
...
+            fluid.default_startup_program().random_seed = seed
+            fluid.default_main_program().random_seed = seed
+
+            exe = fluid.Executor(fluid.CUDAPlace(0))
+
+            resnet = ResNet()
+            optimizer = optimizer_setting(train_parameters)
+
+            np.random.seed(seed)
+            import random
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_imperative_resnet.py]: ...

File: python/paddle/fluid/tests/unittests/test_imperative_resnet.py [hunk 16]
```diff
@@ -0,0 +1,533 @@
...
+            self.assertTrue(np.allclose(value, dy_param_value[key]))
+            self.assertTrue(np.isfinite(value.all()))
+            self.assertFalse(np.isnan(value.any()))
+
+    def test_resnet_cpu_float32(self):
+        seed = 90
+
+        batch_size = train_parameters["batch_size"]
+        batch_num = 1
+        with fluid.imperative.guard(place=fluid.CPUPlace()):
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_imperative_resnet.py]: GPU if available?

File: python/paddle/fluid/tests/unittests/test_imperative_optimizer.py [hunk 17]
```diff
@@ -101,7 +101,7 @@ class TestImperativeMnist(unittest.TestCase):
     def test_mnist_cpu_float32(self):
         seed = 90
 
-        with fluid.imperative.guard():
+        with fluid.imperative.guard(place=fuild.CPUPlace()):
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_imperative_optimizer.py]: GPU if available?

File: python/paddle/fluid/tests/unittests/test_imperative.py [hunk 18]
```diff
@@ -68,15 +68,15 @@ def forward(self, inputs):
 
 class TestImperative(unittest.TestCase):
     def test_layer(self):
-        with fluid.imperative.guard():
+        with fluid.imperative.guard(device=None):
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_imperative.py]: gpu if available?

File: python/paddle/fluid/optimizer.py [hunk 19]
```diff
@@ -381,18 +381,22 @@ def minimize(self,
...
-                        block=loss.block,
-                        name=param._ivar._grad_name(),
-                        stop_gradient=True,
-                        ivar=param._ivar._grad_ivar())
-                    params_grads.append((param, grad_var))
+
+            params_grads = []
+            for param in parameters:
+                if param.stop_gradient:
+                    print("parameter:", param.name, "stop gradient, skip it")
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: don't print?

File: python/paddle/fluid/layers/nn.py [hunk 20]
```diff
@@ -2874,7 +2874,7 @@ def batch_norm(input,
         attr=helper.bias_attr, shape=param_shape, dtype=dtype, is_bias=True)
     # setting stop_gradient=True to reduce computation
     if use_global_stats and helper.bias_attr.learning_rate == 0.:
-        scale.stop_gradient = True
+        bias.stop_gradient = True
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:2877]: Was this a bug or on purpose? @qingqing01

File: python/paddle/fluid/layer_helper.py [hunk 21]
```diff
@@ -435,7 +435,10 @@ def append_activation(self, input_var):
         act_type = act.pop('type')
         tmp = input_var
         # NOTE(dzhwinter): some activation support inplace compution.
-        if not core.IsInplace(act_type):
+        # NOTE(minqiyang): currently, we don't support inplace in imperative mode
+        if not force_no_inplace and core.IsInplace(act_type):
+            tmp = input_var
```
Comment (panyx0718) [inline python/paddle/fluid/layer_helper.py:440]: should we delete this inplace and rely on inplace pass? @dzhwinter

File: python/paddle/fluid/layer_helper.py [hunk 22]
```diff
@@ -419,7 +419,7 @@ def append_bias_op(self, input_var, dim_start=1, dim_end=None):
             attrs={'axis': dim_start})
         return tmp
 
-    def append_activation(self, input_var):
+    def append_activation(self, input_var, force_no_inplace=False):
```
Comment (panyx0718) [inline python/paddle/fluid/layer_helper.py]: no need to add this arg
```
if imperative_mode:
  no_inplace codes
else:
  inplace codes
```

File: python/paddle/fluid/framework.py [hunk 23]
```diff
@@ -67,6 +67,7 @@
 CONTROL_DEP_VAR_PREFIX = core.kControlDepVarName()
 
 _imperative_tracer_ = None
+_current_expected_place_ = None
```
Comment (panyx0718) [inline python/paddle/fluid/framework.py]: `imperative_current_expected_place`?

File: paddle/fluid/pybind/pybind.cc [hunk 24]
```diff
@@ -137,6 +137,14 @@ PYBIND11_MODULE(core, m) {
       .def("_grad_ivar",
            [](const imperative::VarBase &self) { return self.grads_; },
            py::return_value_policy::reference)
+      .def("_to",
```
Comment (panyx0718) [inline paddle/fluid/pybind/pybind.cc]: copy_to?

File: paddle/fluid/imperative/layer.cc [hunk 25]
```diff
@@ -120,6 +168,36 @@ class Autograd {
   }
 };
 
+VarBase* VarBase::NewVarBase(const platform::Place& dst_place,
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.cc]: return a unique pointer and release in pybind?

File: paddle/fluid/imperative/layer.cc (hunk 13 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.cc]: Done

File: python/paddle/fluid/tests/unittests/test_imperative_resnet.py (hunk 14 above)
Comment (velconia) [reply python/paddle/fluid/tests/unittests/test_imperative_resnet.py:219]: done

File: python/paddle/fluid/tests/unittests/test_imperative_resnet.py (hunk 15 above)
Comment (velconia) [reply python/paddle/fluid/tests/unittests/test_imperative_resnet.py]: done

File: python/paddle/fluid/layers/nn.py (hunk 20 above)
Comment (velconia) [reply python/paddle/fluid/layers/nn.py:2877]: @qingqing01 has reviewed this and this is a bug

File: python/paddle/fluid/layer_helper.py (hunk 22 above)
Comment (velconia) [reply python/paddle/fluid/layer_helper.py]: Should keep inplace the same with static graph

File: paddle/fluid/pybind/pybind.cc (hunk 24 above)
Comment (velconia) [reply paddle/fluid/pybind/pybind.cc]: Done

File: python/paddle/fluid/framework.py (hunk 23 above)
Comment (velconia) [reply python/paddle/fluid/framework.py]: Done

File: python/paddle/fluid/tests/unittests/test_imperative_optimizer.py (hunk 17 above)
Comment (velconia) [reply python/paddle/fluid/tests/unittests/test_imperative_optimizer.py]: this is set to CPU on purpose

File: paddle/fluid/imperative/layer.cc (hunk 25 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.cc]: Done

File: python/paddle/fluid/optimizer.py (hunk 19 above)
Comment (velconia) [reply python/paddle/fluid/optimizer.py]: Done

File: paddle/fluid/imperative/layer.cc [hunk 26]
```diff
@@ -120,6 +168,36 @@ class Autograd {
...
+
+    framework::TensorCopySync(var_->Get<framework::LoDTensor>(), dst_place,
+                              tensor);
+
+    dev_ctx->Wait();
+  } else {
+    framework::TensorCopy(var_->Get<framework::LoDTensor>(), dst_place, tensor);
+  }
+
+  if (platform::is_gpu_place(dst_place)) {
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.cc:194]: VLOG_IF

File: paddle/fluid/imperative/layer.h [hunk 27]
```diff
@@ -176,7 +185,8 @@ class OpBase {
       : op_desc_(nullptr),
         forward_id_(-1),
         grad_op_desc_(nullptr),
-        backward_id_(-1) {}
+        backward_id_(-1),
+        place_(platform::CPUPlace()) {}
```
Comment (panyx0718) [inline paddle/fluid/imperative/layer.h:189]: not sure we should default to cpu place...

File: paddle/fluid/imperative/tracer.cc [hunk 28]
```diff
@@ -31,16 +35,38 @@ void CreateGradOp(const framework::OpDesc& op_desc,
...
+      *dev_ctx, grad_var->GetMutable<framework::LoDTensor>(), 0.0);
+}
+
+platform::Place GetExpectedPlace(platform::Place place, VarBasePtrMap inputs) {
+  platform::Place result = place;
+  for (auto it : inputs) {
+    for (VarBase* var : it.second) {
+      platform::Place tmp_place =
+          var->var_->Get<framework::LoDTensor>().place();
+      if (!platform::is_same_place(tmp_place, result)) {
```
Comment (panyx0718) [inline paddle/fluid/imperative/tracer.cc:55]: should it just copy them to the op's device?

File: python/paddle/fluid/imperative/base.py [hunk 29]
```diff
@@ -47,9 +57,7 @@ def to_variable(value, block=None):
             dtype=value.dtype)
         var = py_var._ivar.value()
         tensor = var.get_tensor()
-        tensor.set(value, core.CPUPlace())
+        tensor.set(value, framework._current_expected_place())
         return py_var
     elif isinstance(value, framework.Variable):
         return value
-    else:
```
Comment (panyx0718) [inline python/paddle/fluid/imperative/base.py:54]: ...why remove this...

**Review by panyx0718 — APPROVED** [review]

File: paddle/fluid/imperative/tracer.cc (hunk 28 above)
Comment (velconia) [reply paddle/fluid/imperative/tracer.cc:55]: Good advice, I will add it in next pr

File: paddle/fluid/imperative/layer.h (hunk 27 above)
Comment (velconia) [reply paddle/fluid/imperative/layer.h:189]: Maybe changed to 
```
#ifdef PADDLE_WITH_CUDA
CUDAPlace()
#else 
CPUPlace()
#endif

**Pull request merged**